experimental/FTheoryTools/auxiliary.jl

```diff
@@ -0,0 +1,319 @@
+################################################################
+# 1: Construct auxiliary base space
+################################################################
+
+function _auxiliary_base_space(variable_names::Vector{String}, d::Int)
+    ray_gens = [[if i==j 1 else 0 end for j in 1:length(variable_names)] for i in 1:length(variable_names)]
+    max_cones = Hecke.subsets(length(variable_names), d)
+    auxiliary_base_space = normal_toric_variety(ray_gens, max_cones)
+    set_coordinate_names(auxiliary_base_space, variable_names)
+    return auxiliary_base_space
+end
+
+################################################################
+# 2: Construct ambient space from given base
+################################################################
+
+function _ambient_space_from_base(base::AbstractNormalToricVariety)
+
+    # Extract information about the toric base
+    base_rays = matrix(ZZ, rays(base))
+    base_cones = matrix(ZZ, ray_indices(maximal_cones(base)))
+
+    # Construct the rays for the fibre ambient space
+    xray = [0 for i in 1:ncols(base_rays)+2]
+    yray = [0 for i in 1:ncols(base_rays)+2]
+    zray = [0 for i in 1:ncols(base_rays)+2]
+    xray[ncols(base_rays)+1] = 1
+    yray[ncols(base_rays)+2] = 1
+    zray[ncols(base_rays)+1] = -2
+    zray[ncols(base_rays)+2] = -3
+
+    # Construct the rays of the toric ambient space
+    ambient_space_rays = hcat([r for r in base_rays], [-2 for i in 1:nrows(base_rays)], [-3 for i in 1:nrows(base_rays)])
+    ambient_space_rays = vcat(ambient_space_rays, transpose(xray), transpose(yray), transpose(zray))
+
+    # Construct the incidence matrix for the maximal cones of the ambient space
+    ambient_space_max_cones = []
+    for i in 1:nrows(base_cones)
+        push!(ambient_space_max_cones, [k for k in hcat([b for b in base_cones[i,:]], [1 1 0])])
```

Comment:
WHy is the `k for k` and `b for b` necessary here?

I'd suggest to look at all the `hcat` calls in this PR, I suspect many of them could be optimized

Reply:
I would be happy if it can be optimized. Thing is, I do not know how. Similarly, I do not know how to avoid the `k for k` and `b for b` here. For instance:
```julia
push!(ambient_space_max_cones, hcat(base_cones[i,:], [1 1 0]))
```
gives an `ArgumentError: number of columns of each array must match`.

Best to illustrate with an example what we are actually trying to achieve:

```julia
|julia> base = projective_space(NormalToricVariety,3)
Normal, non-affine, smooth, projective, gorenstein, fano, 3-dimensional toric variety without torusfactor

|julia> base_cones = matrix(ZZ, ray_indices(maximal_cones(base)))
[1   1   1   0]
[0   1   1   1]
[1   0   1   1]
[1   1   0   1]
```
Our goal is to make 3 copies of the first row. To the first copy, we append `[1 1 0]`, to the second copy `[1 0 1]` and to the third `[0 0 1]`. These rows - each consisting are then to be mad the first three rows a matrix. This step we repeat with all other rows of the above matrix. So the desired matrix looks like this:

```julia
[1   1   1   0   1   1   0 ]
[1   1   1   0   1   0   1 ]
[1   1   1   0   0   1   1 ]
[0   1   1   1   1   1   0 ]
...
[1   1   0   1   0   1   1 ]
```
Ultimately, for our purposes, the zeros do not matter (and this is not an allowed input for other functions). So we opt for the incidence matrix of this instead:

```julia
|julia> ambient_space_max_cones = []
Any[]

julia> for i in 1:nrows(base_cones)
           push!(ambient_space_max_cones, [k for k in hcat([b for b in base_cones[i,:]], [1 1 0])])
           push!(ambient_space_max_cones, [k for k in hcat([b for b in base_cones[i,:]], [1 0 1])])
           push!(ambient_space_max_cones, [k for k in hcat([b for b in base_cones[i,:]], [0 1 1])])
       end
julia> ambient_space_max_cones = IncidenceMatrix(vcat(ambient_space_max_cones...))
12×7 IncidenceMatrix
[1, 2, 3, 5, 6]
[1, 2, 3, 5, 7]
[1, 2, 3, 6, 7]
[2, 3, 4, 5, 6]
[2, 3, 4, 5, 7]
[2, 3, 4, 6, 7]
[1, 3, 4, 5, 6]
[1, 3, 4, 5, 7]
[1, 3, 4, 6, 7]
[1, 2, 4, 5, 6]
[1, 2, 4, 5, 7]
[1, 2, 4, 6, 7]
```
(Yes, the very first line is bad since the type is not specified. Yet another thing that can be improved. But one thing at a time.)

Maybe there is a better way to achieve this?